hadoop-ozone/ozonefs/src/main/java/org/apache/hadoop/fs/ozone/BasicRootedOzoneClientAdapterImpl.java

```diff
@@ -0,0 +1,694 @@
+    OFSPath ofsPath = new OFSPath(pathStr);
+    String key = ofsPath.getKeyName();
+    try {
+      OzoneBucket bucket = getBucket(ofsPath, false);
+      OzoneOutputStream ozoneOutputStream = bucket.createFile(
+          key, 0, replicationType, replicationFactor, overWrite, recursive);
+      return new OzoneFSOutputStream(ozoneOutputStream.getOutputStream());
+    } catch (OMException ex) {
+      if (ex.getResult() == OMException.ResultCodes.FILE_ALREADY_EXISTS
+          || ex.getResult() == OMException.ResultCodes.NOT_A_FILE) {
+        throw new FileAlreadyExistsException(
+            ex.getResult().name() + ": " + ex.getMessage());
+      } else {
+        throw ex;
+      }
+    }
+  }
+
+  @Override
+  public void renamePath(String path, String newPath) throws IOException {
+    incrementCounter(Statistic.OBJECTS_RENAMED);
+    OFSPath ofsPath = new OFSPath(path);
+    OFSPath ofsNewPath = new OFSPath(newPath);
+
+    // Check path and newPathName are in the same volume and same bucket.
+    // This should have been checked in BasicRootedOzoneFileSystem#rename
+    // already via regular call path unless bypassed.
+    if (!ofsPath.isInSameBucketAs(ofsNewPath)) {
+      throw new IOException("Cannot rename a key to a different bucket");
+    }
+
+    OzoneBucket bucket = getBucket(ofsPath, false);
+    String key = ofsPath.getKeyName();
+    String newKey = ofsNewPath.getKeyName();
+    bucket.renameKey(key, newKey);
+  }
+
+  /**
+   * Package-private helper function to reduce calls to getBucket().
+   * @param bucket Bucket to operate in.
+   * @param path Existing key path.
+   * @param newPath New key path.
+   * @throws IOException IOException from bucket.renameKey().
+   */
+  void renamePath(OzoneBucket bucket, String path, String newPath)
+      throws IOException {
+    incrementCounter(Statistic.OBJECTS_RENAMED);
+    OFSPath ofsPath = new OFSPath(path);
+    OFSPath ofsNewPath = new OFSPath(newPath);
+    // No same-bucket policy check here since this call path is controlled
+    String key = ofsPath.getKeyName();
+    String newKey = ofsNewPath.getKeyName();
+    bucket.renameKey(key, newKey);
+  }
+
+  /**
+   * Helper method to create an directory specified by key name in bucket.
+   *
+   * @param pathStr path to be created as directory
+   * @return true if the key is created, false otherwise
+   */
+  @Override
+  public boolean createDirectory(String pathStr) throws IOException {
+    LOG.trace("creating dir for path: {}", pathStr);
+    incrementCounter(Statistic.OBJECTS_CREATED);
+    OFSPath ofsPath = new OFSPath(pathStr);
+    // Volume name unspecified, return failure
+    if (ofsPath.getVolumeName().length() == 0) {
+      return false;
+    }
+    // Handle where only volume is specified in pathStr
+    if (ofsPath.getBucketName().length() == 0) {
+      objectStore.createVolume(ofsPath.getVolumeName());
+      return true;
+    }
+    String keyStr = ofsPath.getKeyName();
```

Comment:
what is keyStr is empty(length 0)

Reply:
Actually here we need to put the `keyStr.length()` check after `getBucket`, since `getBucket` need to create volume and bucket if they don't exist. If we check it before `try` clause it wouldn't create the volume and bucket then.

Reply:
Comment updated in c0e2ed7f4cf08490b151613fdb3415b097ee5ac0.